### PR TITLE
G3 collections solution

### PR DESCRIPTION
**Pull request by rosiecakes**
luis rose david

### CODE REVIEW

File: collections_hierarchy/mixins.py [hunk 1]
```diff
@@ -1,30 +1,44 @@
...
 class SequenceMixin(object):
     def __iter__(self):
-        pass
+        data1 = getattr(self, self.DATA_ATTR_NAME)
+        for d in data1:
+            while d:
+                if type(data1) is list:
+                    yield d
+                elif type(data1) is dict:
+                    yield data1[d]
```
Comment (rosiecakes) [inline collections_hierarchy/mixins.py]: this definitely created an infinite loop fml

File: collections_hierarchy/mixins.py [hunk 2]
```diff
@@ -1,54 +1,81 @@
...
+        if isinstance(data1, dict):
+            return data1 == data2
+        else:
+            return sorted(data1) == sorted(data2)
+        
     def __ne__(self, other):
-        # Relies in __eq__
-        pass
-
+        return not self.__eq__(other)
```
Comment (Philipp258) [inline collections_hierarchy/mixins.py:15]: I think 'self == other' is a bit more readable. That's why you define this magic method.

File: collections_hierarchy/mixins.py [hunk 3]
```diff
@@ -1,54 +1,81 @@
...
-        pass
-
+        l = ", ".join([ str(n) for n in self])
+        s = self.DATA_ATTR_NAME
+        
+        if type(getattr(self, s)) is list:
+            return '{}{}{}'.format('[', l, ']')
+        elif type(getattr(self, s)) is dict:
+            return '{}{}{}'.format('{', l, '}')
+
```
Comment (Philipp258) [inline collections_hierarchy/mixins.py:78]: Your variable names make this kind of hard to read.

File: collections_hierarchy/mixins.py [hunk 4]
```diff
@@ -1,54 +1,81 @@
...
-        # Will rely on the iterator
-        # can't do len(self.data)
-        pass
+        # uses the iterator to provide a count of the items
+        c = 0
+        for d in self:
+            c += 1
+        return c
+    
+    count = __len__
```
Comment (Philipp258) [inline collections_hierarchy/mixins.py:47]: Nice, should have thought of that.

File: collections_hierarchy/mixins.py [hunk 5]
```diff
@@ -1,54 +1,81 @@
+from copy import deepcopy
+
 class ComparableMixin(object):
+    #Implements comparison functionality. Common for List and Dict
     def __eq__(self, other):
-        pass
+        data1 = getattr(self, self.DATA_ATTR_NAME) #Get the value of the property that stores the data
+        data2 = getattr(other, other.DATA_ATTR_NAME)
+        
+        if isinstance(data1, dict):
```
Comment (balmf) [inline collections_hierarchy/mixins.py:9]: You want to be careful here as you are not checking that data2 is of the same type as data1. Perhaps something like:

``` python
if isinstance(data1, dict) and isinstance(data2, dict)
```

File: collections_hierarchy/mixins.py [hunk 6]
```diff
@@ -1,54 +1,81 @@
...
+            return data1 == data2
```
Comment (balmf) [inline collections_hierarchy/mixins.py:10]: Also unless I'm missing something here should it not be:

``` python
return sorted(data1) == sorted(data2)
```

for dicts and

``` python
return data1 == data2
```

for the list (the other way around to what you currently have)? As the lists preserve the order of their items you don't want:

``` python
[1,2,3] = [2,1,3]
```

which would return True when sorting the lists before the comparison.

File: collections_hierarchy/mixins.py [hunk 7]
```diff
@@ -60,29 +87,36 @@ def __init__(self, initial=None):
...
 class OperableMixin(object):
     def __add__(self, other):
-        pass
+        new_list = deepcopy(self)
+        new_list.data += other.data
+        return new_list
 
     def __iadd__(self, other):
-        pass
+        return self.__add__(other)
```
Comment (balmf) [inline collections_hierarchy/mixins.py:95]: Calling add will cause a deepcopy to be created which is not entirely necessary of __iadd__. Could instead just do:

``` python
return self.data += other.data
```

but, like __add__ the above does not check that the data is actually stored in the 'data' field of self and other. Need to use `getattr()`

File: collections_hierarchy/mixins.py [hunk 8]
```diff
@@ -60,29 +87,36 @@ def __init__(self, initial=None):
 
 class OperableMixin(object):
     def __add__(self, other):
-        pass
+        new_list = deepcopy(self)
+        new_list.data += other.data
```
Comment (balmf) [inline collections_hierarchy/mixins.py:91]: Assuming data is stored in 'data' field. Ideally would use 

``` python
getattr(self, self.DATA_ATTR_NAME)
```

**Comment by balmf** [comment]
Good code on the whole, easy to read and follow! Few little points to make here and there.
Well done!

**Pull request closed without merging**